core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1502,43 +1492,62 @@ class KafkaApis(val requestChannel: RequestChannel,
       isAcknowledgeDataPresent
     }
 
+    val unauthorizedTopicPartitions: mutable.Set[TopicIdPartition] = mutable.Set.empty[TopicIdPartition]
+    val authorizedTopics = authHelper.filterByAuthorized(
+      request.context,
+      READ,
+      TOPIC,
+      shareFetchData.asScala
+    )(_._1.topicPartition.topic)
+    shareFetchData.forEach((tp, _) => {
+      if (!authorizedTopics.contains(tp.topic))
+        unauthorizedTopicPartitions.add(tp)
+    })
+
     var shareAcknowledgeResponse : ShareAcknowledgeResponse = null
     var shareFetchResponse : ShareFetchResponse = null
 
     // Handling the Acknowledgements from the ShareFetchRequest
-    val isAcknowledgeDataPresent = isAcknowledgeDataPresentInFetchRequest()
-    if (shareSessionEpoch == ShareFetchMetadata.INITIAL_EPOCH) {
-      if (isAcknowledgeDataPresent) {
-        // If this is a full fetch request, it should not contain any acknowledgement Data
-        shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.INVALID_REQUEST.exception) match {
-          case response: ShareAcknowledgeResponse => response
-          case _ => null
+    if (!authHelper.authorize(request.context, READ, GROUP, groupId)) {
+      shareAcknowledgeResponse = shareFetchRequest.getErrorAcknowledgeResponse(AbstractResponse.DEFAULT_THROTTLE_TIME, Errors.GROUP_AUTHORIZATION_FAILED.exception) match {
+        case response: ShareAcknowledgeResponse => response
+        case _ => null
+      }
+    } else {
+      val isAcknowledgeDataPresent = isAcknowledgeDataPresentInFetchRequest()
```

Comment:
Sorry but I don't see this method (`isAcknowledgeDataPresentInFetchRequest`) defined, neither in existing KafkaApis code. What I am missing? Also shouldn't this be first check prior proceeding with any acknowledgement code?

Reply:
I defined it inside the handleShareFetch only.
We don't need to this check at all if for example the authorisation fails for reading the group.